gateway/src/node/client_handling/websocket/connection_handler/ecash/mod.rs

```diff
@@ -0,0 +1,481 @@
+                //binary search returns an arbitrary match, so we have to check for potential multiple matches
+                let mut i = index as i64;
+                while i >= 0 && inner[i as usize].timestamp() == pay_info.timestamp() {
+                    if inner[i as usize] == pay_info {
+                        return Err(RequestHandlingError::InvalidBandwidthCredential(
+                            String::from(
+                                "credential failed to verify on gateway - duplicate payInfo",
+                            ),
+                        ));
+                    }
+                    i -= 1;
+                }
+
+                let mut i = index + 1;
+                while i < inner.len() && inner[i].timestamp() == pay_info.timestamp() {
+                    if inner[i] == pay_info {
+                        return Err(RequestHandlingError::InvalidBandwidthCredential(
+                            String::from(
+                                "credential failed to verify on gateway - duplicate payInfo",
+                            ),
+                        ));
+                    }
+                    i += 1;
+                }
+                Ok(index)
+            }
+        }
+    }
+
+    fn insert_pay_info(&self, pay_info: PayInfo, index: usize) -> Result<(), RequestHandlingError> {
+        let mut inner = self
+            .pay_infos
+            .lock()
+            .map_err(|_| RequestHandlingError::InternalError)?;
+        if index > inner.len() {
+            inner.push(pay_info);
+            return Ok(());
+        }
+        inner.insert(index, pay_info);
+        Ok(())
+    }
+
+    pub async fn check_double_spend(&self, serial_number_bs58: &String) -> bool {
+        if let Some(double_spend_detector) = &self.double_spend_detector {
+            double_spend_detector.check(serial_number_bs58).await
+        } else {
+            false
+        }
+    }
+
+    pub async fn post_credential(
+        &self,
+        api_clients: &[CoconutApiClient],
+        credential: CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        self.cred_sender
+            .unbounded_send(PendingCredential::new(
+                credential.clone(),
+                self.nyxd_client.read().await.address(),
+                api_clients
+                    .iter()
+                    .map(|client| client.api_client.clone())
+                    .collect(),
+            ))
+            .map_err(|_| RequestHandlingError::InternalError)
+    }
+
+    pub async fn spend_online_credential(
+        &self,
+        api_clients: &[CoconutApiClient],
+        credential: &CredentialSpendingRequest,
+    ) -> Result<(), RequestHandlingError> {
+        let serial_number = credential.data.payment.serial_number_bs58();
+        let proposal_id = match credential.data.typ {
+            CredentialType::Voucher => unimplemented!(),
```

Comment:
do we want to do something in that case?

Reply:
If this ever happens, it means you managed to smuggle a coconut bandwidth voucher in a `CredentialSpendingRequest` past a first check. Without code changes, it's impossible to end up in that branch. I'll change that into `unreachable` with a comment

Reply:
TO FIX: scrape backward compatibility